api/dao/storage.py

```diff
@@ -0,0 +1,148 @@
+# @author:  Renzo Frigato
+
+import bson.objectid
+import copy
+import logging
+import re
+
+log = logging.getLogger('scitran.api')
+
+
+class ListStorage(object):
+    """
+    This class provides access to sublists of a mongodb collections elements (called containers).
+    """
+
+    def __init__(self, coll_name, list_name, use_oid = False):
+        self.coll_name = coll_name
+        self.list_name = list_name
+        self.use_oid = use_oid
+        # the collection is not loaded when the class is instantiated
+        # this allows to instantiate the class when the db is not available
+        # dbc is initialized using the load_collection method
+        self.dbc = None
+
+
+    def load_collection(self, db):
+        """
+        Initialize the mongodb collection.
+        """
+        if self.dbc is None:
+            self.dbc = db.get_collection(self.coll_name)
+        return self.dbc
+
+    def get_container(self, _id):
+        """
+        Load a container from the _id.
+
+        This method is usually used to to check permission properties of the container.
+        e.g. list of users that can access the container
+
+        For simplicity we load its full content.
+        """
+        if self.dbc is None:
+            raise RuntimeError('collection not initialized before calling get_container')
+        if self.use_oid:
+            _id = bson.objectid.ObjectId(_id)
+        query = {'_id': _id}
+        log.debug('query {}'.format(query))
+        self.container = self.dbc.find_one(query)
+        return self.container
+
+    def apply_change(self, action, _id, elem_match=None, payload=None):
+        """
+        Generic method to apply an operation.
+        The request is dispatched to the corresponding private methods.
+        """
+        if self.use_oid:
+            _id = bson.objectid.ObjectId(_id)
+        if action == 'GET':
+            return self._get_el(_id, elem_match)
+        if action == 'DELETE':
+            return self._delete_el(_id, elem_match)
+        if action == 'PUT':
+            return self._update_el(_id, elem_match, payload)
+        if action == 'POST':
+            return self._create_el(_id, payload)
+        raise ValueError('action should be one of GET, POST, PUT, DELETE')
+
+    def _create_el(self, _id, payload):
+        log.debug('payload {}'.format(payload))
+        if isinstance(payload, dict):
+            if payload.get('_id') is None:
```

Comment:
the logic for unique items should change, allowing different unique criteria:

e.g. site + uid for permissions, filename for files, no checks for notes